tm2/pkg/sdk/params/handler.go

```diff
@@ -1,79 +1,77 @@
+func NewHandler(params Keeper) paramsHandler {
 	return paramsHandler{
 		params: params,
 	}
 }
 
-func (bh paramsHandler) Process(ctx sdk.Context, msg std.Msg) sdk.Result {
-	errMsg := fmt.Sprintf("unrecognized params message type: %T", msg)
-	return abciResult(std.ErrUnknownRequest(errMsg))
-}
-
-//----------------------------------------
-// Query
+func (ph paramsHandler) Process(_ sdk.Context, msg std.Msg) sdk.Result {
+	errMsg := fmt.Sprintf("%s: %T", errInvalidMsgType, msg)
 
-func (bh paramsHandler) Query(ctx sdk.Context, req abci.RequestQuery) (res abci.ResponseQuery) {
-	switch secondPart(req.Path) {
-	case bh.params.prefix:
-		return bh.queryParam(ctx, req)
-	default:
-		res = sdk.ABCIResponseQueryFromError(
-			std.ErrUnknownRequest("unknown params query endpoint"))
-		return
-	}
+	return sdk.ABCIResultFromError(std.ErrUnknownRequest(errMsg))
 }
 
-// queryParam returns param for a key.
-func (bh paramsHandler) queryParam(ctx sdk.Context, req abci.RequestQuery) (res abci.ResponseQuery) {
-	// parse key from path.
-	key := thirdPartWithSlashes(req.Path)
-	if key == "" {
-		res = sdk.ABCIResponseQueryFromError(
-			std.ErrUnknownRequest("param key is empty"))
+func (ph paramsHandler) Query(ctx sdk.Context, req abci.RequestQuery) abci.ResponseQuery {
+	// Locate the first and second slashes
+	firstSlash, secondSlash := findFirstTwoSlashes(req.Path)
+	if firstSlash == -1 {
+		return sdk.ABCIResponseQueryFromError(
+			std.ErrUnknownRequest(errUnknownQueryEndpoint.Error()),
+		)
 	}
 
-	// XXX: validate?
+	// Extract prefix and key directly from the path
+	var (
+		prefix = req.Path[firstSlash+1 : secondSlash]
+		key    = req.Path[secondSlash+1:]
+	)
 
-	val := bh.params.GetRaw(ctx, key)
+	// Check prefix and key validity
+	if prefix != ph.params.prefix || key == "" {
+		return sdk.ABCIResponseQueryFromError(
+			std.ErrUnknownRequest(errUnknownQueryEndpoint.Error()),
+		)
+	}
 
-	res.Data = val
-	return
-}
+	// Fetch the data and prepare the response
+	var res abci.ResponseQuery
 
-//----------------------------------------
-// misc
+	res.Data = ph.params.GetRaw(ctx, key)
 
-func abciResult(err error) sdk.Result {
-	return sdk.ABCIResultFromError(err)
+	return res
 }
 
-// returns the second component of a path.
-func secondPart(path string) string {
-	parts := strings.SplitN(path, "/", 3)
-	if len(parts) < 2 {
-		return ""
-	} else {
-		return parts[1]
-	}
-}
+// findSlashPositions finds the positions of the first and second slashes in a path.
+// Returns the positions of the slashes and a boolean indicating if both slashes were found
+func findFirstTwoSlashes(s string) (int, int) {
+	first := -1
 
-// returns the third component of a path, including other slashes.
-func thirdPartWithSlashes(path string) string {
-	split := strings.SplitN(path, "/", 3)
-	return split[2]
+	for i := 0; i < len(s); i++ {
+		if s[i] == '/' {
+			if first == -1 {
+				first = i
+			} else {
+				return first, i
+			}
+		}
+	}
+	return -1, -1
 }
```

Comment:
I would suggest something more like this:
```go
func (ph paramsHandler) Query(ctx sdk.Context, req abci.RequestQuery) abci.ResponseQuery {
	// Locate the first
	firstSlash := strings.IndexRune(req.Path, '/')
	if firstSlash < 0 {
		return sdk.ABCIResponseQueryFromError(
			std.ErrUnknownRequest(errUnknownQueryEndpoint.Error()),
		)
	}

	// Extract prefix and key directly from the path
	prefix, key, found := strings.Cut(req.Path[firstSlash+1:], "/")
	if !found || prefix != ph.params.prefix || key == "" {
		return sdk.ABCIResponseQueryFromError(
			std.ErrUnknownRequest(errUnknownQueryEndpoint.Error()),
		)
	}

	// Fetch the data and prepare the response
	var res abci.ResponseQuery
	res.Data = ph.params.GetRaw(ctx, key)
	return res
}
```

Reply:
```shell
goos: darwin
goarch: arm64
pkg: github.com/gnolang/gno/tm2/pkg/sdk/params
cpu: Apple M3 Max
BenchmarkParamsHandler_Query
BenchmarkParamsHandler_Query-14    	11144790	        95.51 ns/op
PASS
```

It's still pretty good 👌 
Definitely better than the original source

Reply:
@petar-dambovaliev, what do you think?

Reply:
> Even if it's slightly less performant, I would suggest something more like this:
> 
> ```go
> func (ph paramsHandler) Query(ctx sdk.Context, req abci.RequestQuery) abci.ResponseQuery {
> 	// Locate the first
> 	firstSlash := strings.IndexRune(req.Path, '/')
> 	if firstSlash < 0 {
> 		return sdk.ABCIResponseQueryFromError(
> 			std.ErrUnknownRequest(errUnknownQueryEndpoint.Error()),
> 		)
> 	}
> 
> 	// Extract prefix and key directly from the path
> 	prefix, key, found := strings.Cut(req.Path[firstSlash+1:], "/")
> 	if !found || prefix != ph.params.prefix || key == "" {
> 		return sdk.ABCIResponseQueryFromError(
> 			std.ErrUnknownRequest(errUnknownQueryEndpoint.Error()),
> 		)
> 	}
> 
> 	// Fetch the data and prepare the response
> 	var res abci.ResponseQuery
> 	res.Data = ph.params.GetRaw(ctx, key)
> 	return res
> }
> ```

What is the reason for this?

Reply:
I was personally thinking it was simpler and more readable, worth the performance loss, but I'm definitely nitpicking here. Feel free to ignore.